### ISSUE TITLE
HDR Support in Web Animation APIs

### ISSUE THREAD

**Issue by mdrejhon** [WebKit/explainers#93]
We are making long-term plans to migrate requestAnimationFrame() systems to other Web Aniumation standards as a low-scientific-quality fallback (see issues in #15) for Apple devices.

We need to know the status of planned HDR support in Web Animation APIs because this will be critical to many future TestUFO HDR tests, including potentially accessing pixel information inside a non-DRM test pattern video file, or direct renders (WebGL, Web Animation API).

**Comment by smfr** [comment]
I don't believe there is any dependency between animation frame rate and HDR.

I'm not sure why you qualify Web Animations as a "low-scientific-quality fallback". They should provide identical callback reliability to requestAnimationFrame.

**Comment by mdrejhon** [comment]
> I don't believe there is any dependency between animation frame rate and HDR.

Thanks; sometimes unintentional happen, like fallbacks (e.g. HDR getting lost) so I wanted to check. 
In this case, this issue can be closed; though kept around as a no-dependency answer;

> I'm not sure why you qualify Web Animations as a "low-scientific-quality fallback". They should provide identical callback reliability to requestAnimationFrame.

We need reliable single-framedrop detection logic (e.g. suddenly know when we see 2 refresh intervals pass instead of 1 refresh interval), to automatically invalidate tests -- some tests require instant notification (e.g. change color away from green) immediately upon a framedrop (any divergence away from framerate=Hz), because the scientific researcher is pointing a camera at the screen, with a mandatory requirement of framerate=Hz.  Sometimes it's a pursuit camera, like this photo:

![image](https://user-images.githubusercontent.com/59981975/180416048-18179cc4-71b6-41df-bbb9-eaae610f5bad.png)

This inexpensive rig outperformed a $30,000 laboratory piece of equipment.

The problem is at first impression, the callback timing data is more opaque.

Although single-framedrop detection is undocumented (it is not part of any web specification), it is relatively reliable in rAF() but not reliable in other animation APIs.

Perhaps in time the APIs will improve; but right now it's still a low-scientific-quality 

I've already explained the considerations elsewhere;

**Comment by smfr** [comment]
Just to be clear, by "HDR" do you mean "high dynamic range", i.e. brighter colors? I'm not sure why you're saying "HDR in Web Animations" since those are two unrelated things.

**Comment by mdrejhon** [comment]
> Just to be clear, by "HDR" do you mean "high dynamic range", i.e. brighter colors? I'm not sure why you're saying "HDR in Web Animations" since those are two unrelated things.

Just wanted to make sure.  Canvas doesn't natively support HDR colorspace (e.g. 32-bit floating point RGB instead of standard classic integer 24-bit RGB), and there's no guarantee that you can pre-multiply to access HDR colors if you try to use third party HDR canvas https://github.com/brianreavis/canvas.hdr.js -- as it might be remapped to a SDR at the OS/driver level, since there's no way to discovery whether the browser is capable of doing it.

**Comment by mdrejhon** [comment]
Please close this item if there's no confirmed HDR dependencies in new Web Animations API

**Comment by mdrejhon** [comment]
Update -- Chromium nightly build supports HDR Canvas.   Might be worth checking if Web Animations API will work with HDR canvases, for all drawing commands, not just drawing just e.g. a HDR AVIF bitmap.   Basically vet all the API to make sure it has an HDR path (e.g. brighter-than-white support)

**Comment by mdrejhon** [comment]
UPDATE:
I also found https://www.w3.org/Graphics/Color/Workshop/ though this is probably different from WHATWG

https://www.w3.org/Graphics/Color/Workshop/slids/Russell.pdf

One thing I noticed is that some WebGL paths use floating-point RGB, including color channel values above 1.0 for colors above SDR, e.g. 2.0 for super-red that is redder than sRGB red.

This is different from the planar RGB that drawing animations of requestAnimationFrame uses, but I will eventually require drawing APIs capable of deeper color, brighter-than-sRGB colors, etc.  There are many ways to theoretically normalize/represent HDR color data.

I realize WHATWG may have a different idea, but regardless, whatever idea is chosen, make sure Web Animations API supports a graceful upgrade path to HDR drawing for things like pixel plotting, line drawing, shape drawing, bitmap sprites (AVIF or HEIF) blending modes, etc.

**Comment by mdrejhon** [comment]
I spent some time to look carefully at possible dependences on Web Animation API, and I don't see any dependency with HDR.

This is because we are preparing to try and support Web Animations API, but we need to make sure we can somehow help influence the API to meet our future needs.  Even requestAnimationFrame() works with HDR canvas, so we may be stuck with it for a decade (or more), but we will be attempting to port to parallel APIs (e.g. WebGL or Web Animations API or other)

Unfortunately, we got more bad news -- Web Animation API is definitely not able to replace requestAnimationFrame() for our needs.

Currently, we're not sure we're able to meet our goals with Web Animation API because of the lack of being able to implement an accurate framedrop-detector (scientific animations worthy of research papers require absolute perfect framerate=Hz perfection) .... 

Our further research still further confirms Web Animation API is lower science-quality than requestAnimationFrame() for our specific needs.  More info why framedrop detection is harder with other APIs than requestAnimationFrame() can seen at https://github.com/WebKit/explainers/issues/93  ...

**Comment by grorg** [comment]
> rAF() is fundamentally a SDR-only animation API.

I don't understand. What does rAF have to do with SDR/HDR? It's just a call to a script. Web Animations is practically the same.

**Comment by mdrejhon** [comment]
> > rAF() is fundamentally a SDR-only animation API.
> 
> I don't understand. What does rAF have to do with SDR/HDR? It's just a call to a script. Web Animations is practically the same.

It's an outdated claim now -- so I take that part back.  It was simply a references to CANVAS being a fundamentally SDR technology, but CANVAS is not necessarily tied to rAF().   After that, I discovered the new Chrome Canary build with the HDR CANVAS support, so I take the statement back.

Initially, I opened this item as a precautionary move, to determine for unintended consequences (very common during standardization).

I've already asked to close this item -- I've now checked much more closely (spent >10 hours) and I can safely say to close this issue.
